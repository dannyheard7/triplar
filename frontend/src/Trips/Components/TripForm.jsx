import React from "react";

import UncontrolledFormFieldGroup from "Forms/Components/UncontrolledFormFieldGroup";
import NonFieldErrors from "Forms/Components/NonFieldErrors";
import DateFieldGroup from "Forms/Components/DateFieldGroup";
import moment from "moment";


export default class  TripForm extends React.Component {
    constructor(props) {
        super(props);

        this.state = {
            startDate: null,
            endDate: null
        };

        this.onDateChange = this.onDateChange.bind(this);
        this.calculateEndDate = this.calculateEndDate.bind(this);
    }

    componentDidMount() {
        this.setState( {
            startDate: this.props.trip ? moment(this.props.trip.startDate) : moment(),
            endDate: this.props.trip ? moment(this.props.trip.endDate) : moment()
        });
    }

    componentDidUpdate(prevProps, prevState) {
        if(this.props.trip && (prevProps.trip.startDate !== this.props.trip.startDate
            || prevProps.trip.endDate !== this.props.trip.endDate)) {
            this.componentDidMount();
        }
    }

    onDateChange(value, date) {
        this.setState({
            [value]: moment(date, "DD/MM/YYYY")
        });
    }

    calculateEndDate(startDate) {
        const trip = this.props.trip ? this.props.trip : {};

        if(trip.endDate) {
            return this.state.endDate ? this.state.endDate : moment();
        } else {
            return startDate.clone().add(3, 'days');
        }
    }

    render() {
        const errors = this.props.errors;
        let trip = (this.props.trip) ? this.props.trip : {};

        let startDate = this.state.startDate ? moment(this.state.startDate) : moment();
        let endDate = this.calculateEndDate(startDate);

        return (
            <form onSubmit={this.props.onSubmit}>
                <NonFieldErrors errors={errors.non_field_errors}/>
                <UncontrolledFormFieldGroup errors={errors.name} label="Trip Name" name="name" type="text"
                                            value={trip.name} required={true}/>
                <DateFieldGroup errors={errors.startDate} label="Start Date" name="startDate"
                                onChange={(date) => this.onDateChange("startDate", date)}
                                value={startDate.format("DD/MM/YYYY")} required={true}/>
                <DateFieldGroup errors={errors.endDate} label="End Date" name="endDate"
                                onChange={(date) => this.onDateChange("endDate", date)}
                                value={endDate.format("DD/MM/YYYY")} required={true}/>
                <button type="submit" className="btn btn-primary">{this.props.submitLabel}</button>
            </form>
        );
    }
}

