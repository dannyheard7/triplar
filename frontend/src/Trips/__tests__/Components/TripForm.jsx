import React from "react";

import TripForm from "Trips/Components/TripForm";
import ShallowRenderer from "react-test-renderer/shallow";

const renderer = new ShallowRenderer();


describe('<TripForm />', () => {
    const props = {
        object: {},
        errors: {}
    };

    test('renders correctly', () => {
        const result = renderer.render(<TripForm {...props} />);
        expect(result).toMatchSnapshot();
    });
});