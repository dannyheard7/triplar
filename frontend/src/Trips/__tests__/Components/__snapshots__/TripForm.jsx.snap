// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<TripForm /> renders correctly 1`] = `
<form>
  <NonFieldErrors />
  <FormFieldGroup
    label="Trip Name"
    name="name"
    required={true}
    type="text"
  />
  <DateFieldGroup
    label="Start Date"
    name="startDate"
    onChange={[Function]}
    required={true}
    value="02/12/2018"
  />
  <DateFieldGroup
    label="End Date"
    name="endDate"
    onChange={[Function]}
    required={true}
    value="05/12/2018"
  />
  <button
    className="btn btn-primary"
    type="submit"
  />
</form>
`;
